### PR TITLE
ORFS Dockerfile incremental optimization

### PR DESCRIPTION
**Pull request by ivysochyn**
# ORFS Dockerfile incremental optimization

This PR splits the final layers in the Docker image to reduce the download size for consecutive pulls, as well as enhances the Docker image build process by optimizing the Dockerfile and leveraging build-cache more effectively.

By separating the image into additional stages, we ensure faster pull speeds when minor changes are introduced, and significantly improve the build process by utilizing previously produced build cache.

You can find example built images under:

* [ghcr.io/antmicro/openroad-flow-scripts-test-cache/ubuntu22.04:latest](https://github.com/antmicro/OpenROAD-flow-scripts/pkgs/container/openroad-flow-scripts-test-cache%2Fubuntu22.04)
* [ghcr.io/antmicro/openroad-flow-scripts-test-cache/ubuntu20.04:latest](https://github.com/antmicro/OpenROAD-flow-scripts/pkgs/container/openroad-flow-scripts-test-cache%2Fubuntu20.04)

CI runs demonstrating the improvements:

1. **Image Built from Scratch**: [View Run](https://github.com/antmicro/OpenROAD-flow-scripts/actions/runs/9890334708)
2. **Minor Modification to README**: This change does not affect the ORFS tools and does not impact most of the ORFS Docker image layers, showcasing the efficiency of the build cache. [View Run](https://github.com/antmicro/OpenROAD-flow-scripts/actions/runs/9891162104) 

This PR introduces the following modifications:

* `docker/Dockerfile.builder`:
  * Multi-Stage Build: The image is divided into multiple stages of copying sources, to leverage caching effectively.
  * Tool Installation: The installation of tools is modified to prevent invalidation of the build cache by avoiding unnecessary copying of git-submodule files.

### CODE REVIEW

File: docker/Dockerfile.builder [hunk 1]
```diff
@@ -1,27 +1,39 @@
...
 
-COPY --from=openroad-flow-scripts-base /OpenROAD-flow-scripts /OpenROAD-flow-scripts
-WORKDIR /OpenROAD-flow-scripts
+RUN echo "" > tools/yosys/abc/.gitcommit && \
+  ./build_openroad.sh --no_init --local --threads ${numThreads}
+
+FROM orfs-base
+
+COPY --link flow/designs flow/designs
+COPY --link flow/platforms flow/platforms
```
Comment (oharboe) [inline docker/Dockerfile.builder]: sort these by descending update size/frequency in git history. Especially platforms is important to have at the top, it is big and rarely changes.

**Review by oharboe — CHANGES_REQUESTED** [review]
suggestion to reduce further. flow/scripts changes freqeuntly and is small and platform changes rarely and is big, so platforms should be before flow/scripts

File: docker/Dockerfile.builder [hunk 2]
```diff
@@ -1,27 +1,39 @@
...
+ARG numThreads=$(nproc)
 
-COPY --from=openroad-flow-scripts-base /OpenROAD-flow-scripts /OpenROAD-flow-scripts
-WORKDIR /OpenROAD-flow-scripts
+RUN echo "" > tools/yosys/abc/.gitcommit && \
+  ./build_openroad.sh --no_init --local --threads ${numThreads}
+
+FROM orfs-base
+
+COPY --link flow/platforms flow/platforms
```
Comment (oharboe) [inline docker/Dockerfile.builder]: add comment that sorting these matter...

File: docker/Dockerfile.builder [hunk 3]
```diff
@@ -1,27 +1,39 @@
...
+COPY --link flow/designs flow/designs
+COPY --link flow/tutorials flow/tutorials
```
Comment (oharboe) [inline docker/Dockerfile.builder:29]: These are still not sorted... Can chatgpt write a small script? :-)

**Review by oharboe — CHANGES_REQUESTED** [review]
document that sorting matters.

File: docker/Dockerfile.builder (hunk 3 above)
Comment (ivysochyn) [reply docker/Dockerfile.builder:29]: As you suggested in a previous comment, those layers are sorted by `update size` (lines)  in descending order.
Let me know if there are any other modifications to the order you want to see.

Here's a look into the statistics for those layers (calculated on this branch starting from the initial commit):
| Name            | Commits | Updated lines    | Docker layer size |
|-----------------|---------|------------|------------|
| flow/platforms  | 519     | 53224454   |  729 MB |
| flow/designs    | 1763    | 11026554   | 284 MB |
| flow/tutorials  | 10      | 36408      | 4.4 MB |
| docs            | 108     | 7146       |  4.1 MB |
| flow/util       | 702     | 18579      |  240 KB |
| flow/scripts    | 839     | 16455      | 112 KB |
| flow/test       | 74      | 291        | 8 KB
| docker          | 9       | 75         | 2 KB

File: docker/Dockerfile.builder (hunk 2 above)
Comment (ivysochyn) [reply docker/Dockerfile.builder]: Added comment stating that layer sorting is performed based on their size.

File: docker/Dockerfile.builder (hunk 3 above)
Comment (oharboe) [reply docker/Dockerfile.builder:29]: You also need to take update frequency into account...

So flow/designs change MUCH more often than tutorials for instance and show be lower down.

Comment (vvbandeira) [reply docker/Dockerfile.builder:29]: @ivysochyn 
Could you share a GH Gitst of the script/command line you used to generate the data? This looks interesting and useful for gathering data from other repos.

Comment (maliberty) [reply docker/Dockerfile.builder:29]: I suspect the rebuilding the <10mb layers is trivial in runtime and designs + platforms is most of the time.  If true, the table order is reasonable.

I expect frequency is most-to-least: scripts, designs, platforms, util, ...

Comment (ivysochyn) [reply docker/Dockerfile.builder:29]: Taking into account both the frequency of commits and the size of the layers, I would rather prioritize frequency. This results in the following order:

| Name            | Commits | Updated lines | Docker layer size |
|-----------------|---------|---------------|-------------------|
| docker          | 9       | 75            | 2 KB              |
| flow/tutorials  | 10      | 36408         | 4.4 MB            |
| docs            | 108     | 7146          | 4.1 MB            |
| flow/test       | 74      | 291           | 8 KB              |
| flow/platforms  | 519     | 53224454      | 729 MB            |
| flow/util       | 702     | 18579         | 240 KB            |
| flow/scripts    | 839     | 16455         | 112 KB            |
| flow/designs    | 1763    | 11026554      | 284 MB            |

This order aims to optimize the cache usage, as the first layers are less likely to be rebuilt frequently.

Additionally, in the changes introduced in this PR, I rely on the [COPY --link](https://docs.docker.com/reference/dockerfile/#copy---link) behavior, which should treat all of those layers independently, preserving the cache even if previous layers are invalidated.

For your reference, here are the tools and commands I used to collect this data.

To calculate the number of commits:
```
git log --oneline <FILE/DIRECTORY> | wc -l
```

Lines updated calculation:
```
git log --shortstat -- <FILE/DIRECTORY> | grep "files changed" | awk '{inserts += $4; deletes += $6} END {print "Insertions: " inserts "\nDeletions: " deletes "\nTotal Changes: " inserts + deletes}
```

To see details of each specific layer of the image, I usually go with [dive](https://github.com/wagoodman/dive):
```
docker run -ti --rm  -v /var/run/docker.sock:/var/run/docker.sock wagoodman/dive <DOCKER_IMAGE_TAG>
```

Comment (oharboe) [reply docker/Dockerfile.builder:29]: Makes sense to.me!

**Review by oharboe — APPROVED** [review]

**Review by vvbandeira — APPROVED** [review]

**Pull request merged**